// MIT License
//
// Copyright (c) 2021 EASE lab
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.vhive.mapreduce";
option java_outer_classname = "mapreduce";
option go_package = "tests/mapreduce/proto";

package mapreduce;

service Mapper {
    rpc Map(MapRequest) returns (MapReply) {}
}

service Reducer {
    rpc Reduce(ReduceRequest) returns (ReduceReply) {}
}

message MapRequest {
    string srcBucket = 1;
    string destBucket = 2;
    repeated Keys keys = 3;
    string jobId = 4;
    int64 mapperId = 5;
    int64 nReducers = 6;
}

message Keys {
    string key = 1;
}

message MapReply {
    string reply = 1;
}

message ReduceRequest {
    string srcBucket = 1;
    string destBucket = 2;
    repeated Keys keys = 3;
    int64 nReducers = 4;
    string jobId = 5;
    int64 reducerId = 6;
}

message ReduceReply {
    string reply = 1;
}

